### PR TITLE
JIT: handle boundary cases for casts of long shifts

### LINKED ISSUE #15291: JITMinOpts=1 Assertion on LSH w/ negative shift
When running Arm64 with JITMinOpts=1, the following test is asserting
JIT/Methodical/flowgraph/dev10_bug679955/indexMinusOne/indexMinusOne.sh

My initial analysis indicates the issue is not platform specific.

```
Assert failure(PID 30534 [0x00007746], Thread: 30534 [0x7746]): Assertion failed 'shiftAmountValue >= 0' in 'ILGEN_0x52ea038a:Method_0xc9efd6b0(int,int,double,char,int,ubyte,double,ubyte,char,byte):char' (IL size 4430)

    File: /home/vmjenkins/workspace/Dotnet/build_and_test/src/jit/morph.cpp Line: 496
    Image: /home/vmjenkins/workspace/Dotnet/build_and_test/bin/tests/pr1/Tests/coreoverlay/corerun
```

When running w/o MinOpts the offending statement is removed.

```
Removing statement [002476] in BB07 as useless:
               [002476] ------------              *  STMT      void  (IL   ???...0x114D)
               [002475] ---XG-------              \--*  RETURN    int   
               [002473] ------------                 |  /--*  CAST      int <- uint <- long
               [002159] ------------                 |  |  |  /--*  CNS_INT   int    -1
               [002472] ------------                 |  |  \--*  LSH       long  
               [002076] ------------                 |  |     \--*  LCL_VAR   long   V128 tmp103      
               [002474] ---XG-------                 \--*  INDEX     int   
               [001907] ------------                    \--*  LCL_VAR   ref    V119 tmp94       
```

With Minopts, the statement remains and the above assert fires.

Offending code is as follows
```
                if (shiftAmount->IsIntegralConst())
                {
                    const ssize_t shiftAmountValue = shiftAmount->AsIntCon()->IconValue();
                    assert(shiftAmountValue >= 0);

                    if (shiftAmountValue >= 32)
                    {
                        // Result of the shift is zero.
                        DEBUG_DESTROY_NODE(tree);
                        GenTree* zero = gtNewZeroConNode(TYP_INT);
                        return fgMorphTree(zero);
                    }
                    else
                    {
                        // Shift amount is small enough that we can push the cast through.
                        canPushCast = true;
                    }
                }
                else
                {
                    // Shift amount is unknown. We can't optimize this case.
                    assert(!canPushCast);
                }
```

The assert could be removed and the negative case could be properly handled.  Whichever is correct:
+ Treat negative shift as `canPushCast = false;`
+ Calculate the effective shift, which I suspect is `actualShift = shiftAmount & (bitsInType(T) - 1)`

Otherwise
+ Ignore the MinOpts failure on illegal IL.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Remove the assert that the shift count is non-negative.
Don't try and optimize if the shift count is >= 64 or < 0.
Update test case to cover these values.

Updates the fix from #15236.
Closes #15291.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@briansull PTAL
cc @dotnet/jit-contrib 

Note CSC will `and` the shift count with 63 (for longs) so the new test cases don't actually cover the extreme values. I suppose I can add an IL test for these. Hopefully the undefined behavior is not platform dependent.

**Comment by BruceForstall** [comment]
fyi, I just saw this in armlb run:
```
BEGIN EXECUTION
 "C:\Users\robox\j\workspace\armlb_cross_c---052e33bb\bin\tests\Windows_NT.armlb.Checked\Tests\Core_Root\corerun.exe" indexMinusOne.exe 

Assert failure(PID 17912 [0x000045f8], Thread: 14332 [0x37fc]): Assertion failed 'shiftAmountValue >= 0' in 'ILGEN_0x52ea038a:Method_0xc9efd6b0(int,int,double,char,int,ubyte,double,ubyte,char,byte):char' (IL size 4430)

    File: d:\j\workspace\armlb_cross_c---17d9d8e9\src\jit\morph.cpp Line: 496
    Image: C:\Users\robox\j\workspace\armlb_cross_c---052e33bb\bin\tests\Windows_NT.armlb.Checked\Tests\Core_Root\CoreRun.exe

Expected: 100
Actual: 123456789
END EXECUTION - FAILED
FAILED
```
https://ci.dot.net/job/dotnet_coreclr/job/master/job/armlb_cross_checked_windows_nt_tst_prtest/198/

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT armlb Cross Checked Innerloop Build and Test

**Comment by sandreenko** [comment]
Checked that this PR solves recent ILGen regressions.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -493,20 +493,31 @@ GenTreePtr Compiler::fgMorphCast(GenTreePtr tree)
                 if (shiftAmount->IsIntegralConst())
                 {
                     const ssize_t shiftAmountValue = shiftAmount->AsIntCon()->IconValue();
-                    assert(shiftAmountValue >= 0);
 
-                    if (shiftAmountValue >= 32)
+                    if (shiftAmountValue >= 64)
```
Comment (BruceForstall) [inline src/jit/morph.cpp]: Isn't this value dependent on architecture? So for x86, >=32, but for x64 >= 64?

Comment (BruceForstall) [reply src/jit/morph.cpp]: Or, not by architecture, but type of value being shifted. From the CLI spec:
```
Stack Transition:
…, value, shiftAmount => …, result
Description:
The shl instruction shifts value (int32, int64 or native int) left by the number of bits specified by shiftAmount. shiftAmount is of type int32 or native int. The return value is unspecified if shiftAmount is greater than or equal to the width of value.
```

Comment (AndyAyersMS) [reply src/jit/morph.cpp]: This code is only reachable when we have a long shift.

Comment (mikedn) [reply src/jit/morph.cpp]: As far as I can tell the whole code is under `if ((srcType == TYP_LONG) && ((dstType == TYP_INT) || (dstType == TYP_UINT)))` so we're always dealing with a 64 bit shift.

**Comment by AndyAyersMS** [comment]
@briansull not sure if you saw this ... would like you to approve too.

**Pull request merged**